### ISSUE TITLE
possible optimization for slice / concatenate operations over contiguous chunks

### ISSUE THREAD

**Issue by rabernat** [dask/dask#3535]
Dask's [`array.roll`](https://github.com/dask/dask/blob/4020cb97a1d356d0b05c9c9b5c96f1aa277f3bf9/dask/array/routines.py#L875-L920) function shifts the positions of array elements along an axis, wrapping around the edge of the array. This operation is very important for many numerical calculations. It does this by slicing the array and then calling `concatenate` to piece it back together in the new order.

Xarray's [`roll` method](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.roll.html) basically [re-implements](https://github.com/pydata/xarray/blob/master/xarray/core/variable.py#L985-L1003) the same logic. In [xgcm](http://xgcm.readthedocs.io/en/latest/), we we also do something similar, but possibly using more complex topological connections between elements of the array than the simple periodicity assumed by `roll`. I am motivated to raise this issue here because of scalability challenges related to xgcm on very large datasets.

I think there is a opportunity for optimization of these operations in the **special case where the roll occurs over a contiguous chunk**. (In practice, this is often the case.) Consider the following example.

```python
import numpy as np
import dask.array as dsa

nt = 2
nx = 500
data = dsa.random.random(size=(nt, nx), chunks=(1, nx))
dsa.roll(data, 1, axis=-1).mean().visualize()
```
![image](https://user-images.githubusercontent.com/1197350/40553065-8f80d2aa-600f-11e8-9d14-8ec40ea46a8f.png)

It is instructive to re-implement our own roll function the way dask does it to see that it works the same:

```python
def roll_with_dask(array, split_point):
    left = array[..., :split_point]
    right = array[..., split_point:]
    return dsa.concatenate([right, left], axis=-1)
roll_with_dask(data, 1).mean().visualize()
```
![image](https://user-images.githubusercontent.com/1197350/40553059-862b2412-600f-11e8-83ae-fc6bdd2f45a9.png)

Now I implement the same function in numpy and apply it with map blocks:
```python
def roll_with_numpy(array, split_point):
    left = array[..., :split_point]
    right = array[..., split_point:]
    return np.concatenate([right, left], axis=-1)
data.map_blocks(roll_with_numpy, 1).visualize()
```
![image](https://user-images.githubusercontent.com/1197350/40553157-ce6241d4-600f-11e8-8753-029274ff8039.png)

These two give the same answer:
```python
np.testing.assert_allclose(
    data.map_blocks(roll_with_numpy, 1),
    roll_with_dask(data, 1)
)
```

But the `map_blocks` version has fewer chunks and consequently is much more efficient. Furthermore, no rechunking is required down the line if we want to combine this rolled array with other non-rolled array (e.g. when taking a finite difference). These things make a big difference when working with very large arrays.

To generalize this example, it seems that all slice and concat operations over contiguous chunks should be able to be accomplished at the numpy level. Is this something that could conceivably be detected at the optimization stage and in-lined?

**Comment by jakirkham** [comment]
IIUC we want `roll` to not split a chunk if that chunk fills the axis being rolled, but instead we want it to perform the `roll` operation within the chunks to keep the chunks intact? If so, that sounds reasonable. Any interest in doing a PR?

**Comment by rabernat** [comment]
> IIUC we want roll to not split a chunk if that chunk fills the axis being rolled, but instead we want it to perform the roll operation within the chunks to keep the chunks intact? If so, that sounds reasonable. Any interest in doing a PR?

Yes this is what I meant. But I was suggesting something more ambitious that just modifying roll; I am suggesting to automatically detect such cases where the operations all occur on contiguous chunks and then fusing them into a single task. I recognize this may be too vague and complex to be actionable.

Updating roll will not help xarray or xgcm, which instead implement their own slightly different but fundamentally similar versions of roll. But automatic fusing of such operations would.

**Comment by mrocklin** [comment]
Using map_blocks when we roll over single-chunked dimensions cdwould be a welcome change.

If you're looking for fusion then you probably want to set `dask.config.set(fuse_ave_width=5)` or something similar.
